model/config_test.go

```diff
@@ -499,3 +507,384 @@ func TestKeyString(t *testing.T) {
+				Endpoint: endpoint1,
+			},
+		},
+		{
+			name: "invalid service",
+			instance: &ServiceInstance{
+				Service: &Service{},
+			},
+		},
+		{
+			name: "invalid endpoint port and service port",
+			instance: &ServiceInstance{
+				Service: service1,
+				Endpoint: NetworkEndpoint{
+					Address: "192.168.1.2",
+					Port:    -80,
+				},
+			},
+		},
+		{
+			name: "endpoint missing service port",
+			instance: &ServiceInstance{
+				Service: service1,
+				Endpoint: NetworkEndpoint{
+					Address: "192.168.1.2",
+					Port:    service1.Ports[1].Port,
+					ServicePort: &Port{
+						Name:     service1.Ports[1].Name + "-extra",
+						Port:     service1.Ports[1].Port,
+						Protocol: service1.Ports[1].Protocol,
+					},
+				},
+			},
+		},
+		{
+			name: "endpoint port and protocol mismatch",
+			instance: &ServiceInstance{
+				Service: service1,
+				Endpoint: NetworkEndpoint{
+					Address: "192.168.1.2",
+					Port:    service1.Ports[1].Port,
+					ServicePort: &Port{
+						Name:     "http",
+						Port:     service1.Ports[1].Port + 1,
+						Protocol: ProtocolGRPC,
+					},
+				},
+			},
+		},
+	}
+	for _, c := range cases {
+		if got := c.instance.Validate(); (got == nil) != c.valid {
+			t.Errorf("%s failed: got valid=%v but wanted valid=%v: %v", c.name, got == nil, c.valid, got)
+		}
+	}
+}
+
+func TestServiceValidate(t *testing.T) {
+	ports := PortList{
+		{Name: "http", Port: 80, Protocol: ProtocolHTTP},
+		{Name: "http-alt", Port: 8080, Protocol: ProtocolHTTP},
+	}
+	badPorts := PortList{
+		{Port: 80, Protocol: ProtocolHTTP},
+		{Name: "http-alt^", Port: 8080, Protocol: ProtocolHTTP},
+		{Name: "http", Port: -80, Protocol: ProtocolHTTP},
+	}
+
+	address := "192.168.1.1"
+
+	cases := []struct {
+		name    string
+		service *Service
+		valid   bool
+	}{
+		{
+			name:    "empty hostname",
+			service: &Service{Hostname: "", Address: address, Ports: ports},
+		},
+		{
+			name:    "invalid hostname",
+			service: &Service{Hostname: "hostname.^.com", Address: address, Ports: ports},
+		},
+		{
+			name:    "empty ports",
+			service: &Service{Hostname: "hostname", Address: address},
+		},
+		{
+			name:    "bad ports",
+			service: &Service{Hostname: "hostname", Address: address, Ports: badPorts},
+		},
+	}
+	for _, c := range cases {
+		if got := c.service.Validate(); (got == nil) != c.valid {
+			t.Errorf("%s failed: got valid=%v but wanted valid=%v: %v", c.name, got == nil, c.valid, got)
+		}
+	}
+}
+
+func TestTagsValidate(t *testing.T) {
+	cases := []struct {
+		name  string
+		tags  Tags
+		valid bool
+	}{
+		{
+			name: "empty tags",
+		},
+		{
+			name: "bad tag",
+			tags: Tags{"^": "^"},
+		},
+		{
+			name:  "good tag",
+			tags:  Tags{"key": "value"},
+			valid: true,
+		},
+	}
+	for _, c := range cases {
+		if got := c.tags.Validate(); (got == nil) != c.valid {
+			t.Errorf("%s failed: got valid=%v but wanted valid=%v: %v", c.name, got == nil, c.valid, got)
+		}
+	}
+}
+
+func TestTestEquals(t *testing.T) {
```

Comment:
Typo? There are some tests in `service_test.go`, maybe worth combining them together in one file?